### ISSUE TITLE
embedsMany does not populate auto id

### ISSUE THREAD

**Issue by james-gardner** [strongloop/loopback#1552]
There are a couple of issues relating to errors being thrown relating to embedsMany and auto id:

https://github.com/strongloop/loopback/issues/556
https://github.com/strongloop/loopback-datasource-juggler/issues/336

Both of which are closed. I am however still able to reproduce the index issue and get the following message:

    {
      "error": {
        "name": "ValidationError",
        "status": 422,
        "message": "The `order` instance is not valid. Details: `items` contains invalid item at index `0`: `id` is blank (value: [ { name: 'Sample Product...} ]).",
        "statusCode": 422,
        "details": {
          "context": "order",
          "codes": {
            "items": [
              "invalid"
            ]
          },
          "messages": {
            "items": [
              "contains invalid item at index `0`: `id` is blank"
            ]
          }
        }
      }
    }

I've defined models for 'customer', 'order' and 'items'.

Customer

    {
        "name": "customer",
        "base": "PersistedModel",
        "idInjection": true,
        "properties": {
            "firstName": {
                "type": "string",
                "required": true
            },
            "lastName": {
                "type": "string",
                "required": true
            }
        },
        "validations": [],
        "relations": {
            "order": {
                "type": "hasMany",
                "model": "order",
                "foreignKey": ""
            }
        },
        "acls": [],
        "methods": []
    }

Order

    {
        "name": "order",
        "base": "PersistedModel",
        "idInjection": true,
        "properties": {},
        "validations": [],
        "relations": {
            "customer": {
                "type": "belongsTo",
                "model": "customer"
            },
            "item": {
                "type": "embedsMany",
                "model": "item",
                "options": {
                    "validate": true,
                    "forceId": false
                }
            }
        },
        "acls": [],
        "methods": []
    }

Item

    {
        "name": "item",
        "base": "PersistedModel",
        "idInjection": true,
        "properties": {
            "name": {
                "type": "string",
                "required": true
            },
            "description": {
                "type": "string",
                "required": true
            },
            "quantity": {
                "type": "number",
                "required": true
            },
            "price": {
                "type": "number",
                "required": true
            }
        },
        "validations": [],
        "acls": [],
        "methods": []
    }

I get the error when I try and POST the following data to 'order':

    {
        "customerId": 1,
        "items": [
            {   
                "name": "Sample Product A",
                "description": "An exotic and rare example product crafted with indulgence.",
                "quantity": 1,
                "price": 4.99,
                "orderId": 1
            }
        ]
    }

**Comment by satyadeepk** [comment]
I am seeing the same issue as well, when using:
```
"options": {
                "validate": true,
                "forceId": false
            }
```
Have you found any workaround?

**Comment by jamesjjk** [comment]
No this is still not fixed even if you POST/PUT multiple entities to embedsmany through a related model, with all permutations of options to a transient model it will fail. This needs to be fixed has been an issue for a long time and doesn't feel production ready.

**Comment by raymondfeng** [comment]
@fabien Do you have some ideas?

**Comment by jamesjjk** [comment]
@raymondfeng Apologies should have added the error: When submitting a POST of a related model which embedsmany X. It simply does not return or generate an ID for the embedded transient model. If submitting an array of embedded items: instance is not valid. Details: `List` contains duplicate `id`...

Hope it helps, but it seems to be the same problem throughout multiple threads..

**Comment by fabien** [comment]
@james-gardner is your Item model using a `transient` data source? It appears that it's using the same `PersistedModel` setup as the rest of your models. It might be helpful to pinpoint the issue.

In any case, there are quite extensive tests to cover the different `embedsMany` setups with id handling:
- https://github.com/strongloop/loopback-datasource-juggler/blob/master/test/relations.test.js#L4007
- https://github.com/strongloop/loopback-datasource-juggler/blob/master/test/relations.test.js#L3968
- https://github.com/strongloop/loopback-datasource-juggler/blob/master/test/relations.test.js#L4007
- https://github.com/strongloop/loopback-datasource-juggler/blob/master/test/relations.test.js#L4178

If anyone can create a basic test app repo with a failing setup, I'll look into it.

**Comment by fabien** [comment]
On second thought, here's what's happening: 

`embedsMany` should be treated as an actual relation, no different from `hasMany`.

This means that you cannot just POST the full object with embedded/nested data to create everything all at once. This has never been possible, and if it would have been, it should be a feature that then would also work with `hasMany` for example. Note that this is only an issue on `create`, but you'll want to treat updates in the same manner, for example: `PUT /orders/:id/items/:pk`.

In other words, the only right way to create a new order at the moment would be:
- POST `/orders' only with `customerId`
- POST `/orders/:id/items` with your item data

@raymondfeng, I believe this would be an area to improve upon: handle nested data for relations, and ideally use the connector's ability to do this atomically where possible ($push, $pull on MongoDB for example).

**Comment by fabien** [comment]
@crandmck this would be a good thing to point out in the docs. Thanks!

**Comment by jamesjjk** [comment]
@fabien Hi Fabien, I understand - that said it still has a problem return an ID for a nested object, but it does create the single record - furthermore the explorer implies an array of objects for a nested component. For example Customer/:id/order/:id/items. When adding a single item (as you said it doesn't support multiple nested items) - where items is a transient source, it does not return the ID. Whilst it does should you create the item under order directly /order/:id/items

Hope that makes sense.

**Comment by jamesjjk** [comment]
@fabien Also, I just checked and the hasMany does support nested data through the API. EmbedsMany throws an error, for example:
/items
[{
  "name": "book",
  "price": 5
},{
  "name": "chair",
  "price": 4
}]

ValidationError: The `Item` instance is not valid. Details: `name` can't be blank (value: undefined); `price` can't be blank (value: undefined).\n

**Comment by jamesjjk** [comment]
It seems this is the same case for referencesMany, which also does not supported nested arrays of items.

**Comment by jamesjjk** [comment]
ReferencesMany also ignores the 'property' name when set..

**Comment by fabien** [comment]
@jamesjjk regarding `hasMany` with nested data, which REST endpoint are you using? I'm pretty sure it is not supported at `/api/orders/id:/items`, and this is why there are no tests for it, nor documentation.

It probably works because it falls back to normal `create` which does support arrays. But that's not really intended behavior, it's a coincidence. Although I can see the value of having this.

Regarding the creation of a single item, I have extensive (integration) test cases in my app(s), with quite some `embedsMany` relations, and I have yet to experience your issue of not getting back an ID, even with Transient datasource. Do you have more details?

**Comment by fabien** [comment]
PS Customer/:id/order/:id/items is a double-nested relation, which is not supported unless you use used `nestRemoting` explicitly. So I'm curious how you achieved such an endpoint ;)

**Comment by jamesjjk** [comment]
@fabien Here is what I did Customer [hasMany] > Order [EmbedsMany] items. I understand the create and upsert method support array, but it doesn't seem to when it comes to Embedmany and references many. In your tests you have a test which tries to create 'items' (embedded) using an array which fails due to validation of a missing property. With me the test fails anyways because its not accepting an array? Any ideas?

**Comment by jamesjjk** [comment]
@fabien On further inspection it seems embedded (embedsMany) models defined in JSON do not pass the test they do not implement or inherit the Model properties. These tests fail: https://github.com/strongloop/loopback-datasource-juggler/blob/master/test/relations.test.js#L3749

**Comment by fabien** [comment]
@jamesjjk I'm a bit puzzled, is this test case failing as-is when your run the Juggler test suite? Or do you see this fail in your own app with your own code? There should be no difference in how you define your models, either in code or in JSON.

How did you setup your models exactly? Can you post `customer.json/js`, `order.json/js`, `item.json/js` and `model-config.json` alongside `datasources.json`?

**Comment by jamesjjk** [comment]
I have attached the models in question. Please note that the response from
create action does not inherit the model props:

Order.postOrder = function(data, cb) {

    let _data = {
      'foobar': 5,
      'productId': 12345
    }

    Order.create(_data, function(err, _res) {
      if(err){
        console.log(err);
      }
      //DOES NOT INHERIT MODEL CLASS LIKE HASMANY ETC.
      console.log(_res.itemList);
    });
  }

**Comment by jamesjjk** [comment]
@fabien now with files:

datasources:

{
  "db": {
    "name": "db",
    "connector": "memory"
  },
  "transient": {
    "name": "transient",
    "connector": "transient"
  },
  "mongodb": {
    "name": "mongodb",
    "connector": "mongodb"
  }
}

order.json

{
  "name": "Order",
  "base": "PersistedModel",
  "idInjection": true,
  "mixins": {
    "TimeStamp": true
  },
  "options": {
    "validateUpsert": false
  },
  "properties": {
    "foobar": {
      "type": "number",
      "required": true
    }
  },
  "validations": [],
  "relations": {
    "product": {
      "type": "belongsTo",
      "model": "Product",
      "foreignKey": "productId"
    },
    "item": {
      "type": "embedsMany",
      "model": "Item",
      "property": "itemList",
      "options": {
        "validate": true,
        "forceId": false
      }
    }
  },
  "acls": [],
  "methods": []
}

item.json

{
  "name": "Item",
  "base": "Model",
  "idInjection": false,
  "properties": {
    "someId": {
      "type": "number",
      "required": true,
      "id": true
    },
    "price": {
      "type": "number",
      "required": true
    }
  },
  "validations": [],
  "acls": [],
  "methods": []
}

ModelConfig:
  "Order": {
    "dataSource": "mongodb",
    "public": true
  },
  "Item": {
    "dataSource": "transient",
    "public": false
  }

Order.js


module.exports = function(Order) {

  Order.postOrder = function(data, cb) {

    let _data = {
      'foobar': 5,
      'productId': 12345
    }

    Order.create(_data, function(err, _res) {
      if(err){
        console.log(err);
      }
      //DOES NOT INHERIT MODEL CLASS LIKE HASMANY ETC.
      console.log(_res.itemList);
    });
  }

  Order.remoteMethod(
      'postOrder',
      {
        accepts: {arg: 'data', type: 'object', http: { source: 'body' }},
        http: {path: '/postOrder', verb: 'post'},
        returns: {arg: 'result', type: 'Array'}
      }
  );


};

**Comment by jamesjjk** [comment]
@fabien I have figured out what I did wrong - I should have been using the embedded models name not the property to reference it a function. This is my mistake, I appologies for the inconvenience. Still an issue remaining with the IDs when 'It probably works because it falls back to normal create which does support arrays. But that's not really intended behavior, it's a coincidence. Although I can see the value of having this functionality.' But as you said that is not the intended behavior. Thanks for all your help!

**Comment by fabien** [comment]
@jamesjjk I'm glad that you have sorted it out! Regarding array handling, please feel free to open a separate issue for this. In that case we'll aim to unify this across all relations with multiple return values (toMany, ...)

**Comment by crandmck** [comment]
> this would be a good thing to point out in the docs. Thanks!

@fabien It's a bit hard for me to sort out exactly what should be documented here... I _think_ I have it.  PTAL: http://docs.strongloop.com/display/LB/Embedded+models+and+relations#Embeddedmodelsandrelations-EmbedsMany

**Comment by fabien** [comment]
@crandmck LGTM, thanks!
